cpp/include/cudf/strings/string_view.cuh

```diff
@@ -322,31 +399,25 @@ __host__ __device__ size_type to_char_utf8(const char* str, char_utf8& character
  * @param[out] str Allocated char array with enough space to hold the encoded characer.
  * @return The number of bytes in the character
  */
-__host__ __device__ size_type from_char_utf8(char_utf8 character, char* str);
-
-/**
- * @brief Return the number of UTF-8 characters in this provided char array.
- *
- * @param str String with encoded char bytes.
- * @param bytes Number of bytes in str.
- * @return The number of characters in the array.
- */
-__host__ __device__ size_type characters_in_string(const char* str, size_type bytes);
-
-/**
- * @brief This will return true if passed the first byte of a UTF-8 character.
- *
- * @param byte Any byte from a valid UTF-8 character
- * @return true if this the first byte of the character
- */
-constexpr bool is_begin_utf8_char(uint8_t byte)
+__host__ __device__ inline size_type from_char_utf8(char_utf8 character, char* str)
 {
-  // The (0xC0 & 0x80) bit pattern identifies a continuation byte of a character.
-  return (byte & 0xC0) != 0x80;
+  size_type const chr_width = bytes_in_char_utf8(character);
+  for (size_type idx = 0; idx < chr_width; ++idx) {
+    str[chr_width - idx - 1] = static_cast<char>(character) & 0xFF;
+    character                = character >> 8;
+  }
+  return chr_width;
 }
 
 }  // namespace detail
 }  // namespace strings
 }  // namespace cudf
 
+// Using device specific intrinsic functions in device only code causes a
+// compile error for .cpp files.
+// Many .cpp files need only the host member functions defined above.
+// Therefore, the device only functions are isolated in the .inl file
+// included below.
+#if defined(__CUDA_ARCH__)
```

Comment:
`__CUDA_ARCH__` is only valid when used inside a function. Using it at namespace scope like this is UB.

Reply:
I need something like this to prevent including the `.inl` file when compiled by the host compiler and to only include the `.inl` when compiled by nvcc.
Otherwise, I could add it to every `__device__` function inside the `.inl` file instead.

Reply:
> `__CUDA_ARCH__` is only valid when used inside a function. Using it at namespace scope like this is UB.

[CUDA Documentation](https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#application-compatibility
) says `It is only defined for device code`. 
To be specific, it's device compilation phase. `when __CUDA_ARCH__ is defined (i.e., during device compilation phase)` 

`__CUDA_ARCH__` restrictions https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#cuda-arch-macro

nvcc can be identified using `__CUDACC__`. https://docs.nvidia.com/cuda/cuda-compiler-driver-nvcc/index.html#nvcc-identification-macro 
https://stackoverflow.com/a/8809924/1550940

To summarize,
`__CUDA_ARCH__` is defined during device compilation phase (for nvcc host compilation phase, it's not defined).
`__CUDACC__` is defined only when nvcc is compiling CUDA source files. (for C/C++ files it's not defined). But it is defined for both device compilation and host compilation phase.

Reply:
> __CUDACC__ is defined only when nvcc is compiling CUDA source files
This is exactly what I need.